### PR TITLE
feat(pie-cookie-banner): DSW-1127 add manage preferences modal

### PR DESCRIPTION
**Pull request by jamieomaguire**
---
"@justeattakeaway/pie-cookie-banner": minor
---

[Added] - Add a manage preferences modal to the cookie banner and event emitters to notify consumers of user interactions

---
"@justeattakeaway/pie-modal": minor
---

[Added] - leading and supporting action click events to notify consumers

---

demo

### CODE REVIEW

File: packages/components/pie-cookie-banner/src/index.ts [hunk 1]
```diff
@@ -1,17 +1,106 @@
...
+ * @event {CustomEvent} pie-cookie-banner-accept-all - when all cookies are accepted.
+ * @event {CustomEvent} pie-cookie-banner-necessary-only - when all only necessary cookies are accepted.
+ * @event {CustomEvent} pie-cookie-manage-prefs - when a user clicks manage preferences.
+ */
 export class PieCookieBanner extends LitElement implements CookieBannerProps {
+    @state()
+    private _isCookieBannerHidden = false;
+
+    @state()
+    private _isModalOpen = false;
```
Comment (raoufswe) [inline packages/components/pie-cookie-banner/src/index.ts:28]: i think this is the first import of a lit state in the codebase 🥇 😆

**Review by raoufswe — DISMISSED** [review]

**Review by raoufswe — DISMISSED** [review]

**Review by raoufswe — APPROVED** [review]

**Review by raoufswe — APPROVED** [review]

File: packages/components/pie-cookie-banner/test/component/pie-cookie-banner.spec.ts [hunk 2]
```diff
@@ -17,4 +41,112 @@ test.describe('PieCookieBanner - Component tests', () => {
...
+    test('clicking save in "Manage preferences" closes the modal and closes the cookie banner', async ({ mount, page }) => {
+        // Arrange
+        await mount(PieCookieBanner, {
+            props: {} as CookieBannerProps,
+        });
+
+        const cookieBanner = await page.locator(componentSelector);
+
+        // Act
+        await page.locator(managePreferencesSelector).click();
```
Comment (xander-marjoram) [inline packages/components/pie-cookie-banner/test/component/pie-cookie-banner.spec.ts:143]: Not requesting changes, but I think you can do `await page.click(selector)` 😄

Comment (jamieomaguire) [reply packages/components/pie-cookie-banner/test/component/pie-cookie-banner.spec.ts:143]: perfect! I'll update in the follow up PR if ok :)

**Review by xander-marjoram — APPROVED** [review]

**Review by ashleynolan — APPROVED** [review]

File: packages/components/pie-cookie-banner/test/component/pie-cookie-banner.spec.ts [hunk 3]
```diff
@@ -17,4 +41,112 @@ test.describe('PieCookieBanner - Component tests', () => {
...
+            },
+        });
+
+        const cookieBanner = await page.locator(componentSelector);
+
+        // Act
+        await page.locator(acceptAllSelector).click();
+
+        // Assert
+        expect(events).toHaveLength(1);
```
Comment (kevinrodrigues) [inline packages/components/pie-cookie-banner/test/component/pie-cookie-banner.spec.ts:63]: Does the name not get picked up? I think i had a similar issue somewhere before but wondered if it was the same thing here too now?

File: packages/components/pie-cookie-banner/test/component/pie-cookie-banner.spec.ts [hunk 4]
```diff
@@ -17,4 +41,112 @@ test.describe('PieCookieBanner - Component tests', () => {
...
+
+        await mount(PieCookieBanner, {
+            props: {} as CookieBannerProps,
+
+            on: {
+                [ON_COOKIE_BANNER_ACCEPT_ALL]: (event: Event) => events.push(event),
+            },
+        });
+
+        const cookieBanner = await page.locator(componentSelector);
```
Comment (kevinrodrigues) [inline packages/components/pie-cookie-banner/test/component/pie-cookie-banner.spec.ts:57]: I dont think you need the `await` here.

File: packages/components/pie-cookie-banner/test/component/pie-cookie-banner.spec.ts [hunk 5]
```diff
@@ -17,4 +41,112 @@ test.describe('PieCookieBanner - Component tests', () => {
...
+
+        await mount(PieCookieBanner, {
+            props: {} as CookieBannerProps,
+
+            on: {
+                [ON_COOKIE_BANNER_NECESSARY_ONLY]: (event: Event) => events.push(event),
+            },
+        });
+
+        const cookieBanner = await page.locator(componentSelector);
```
Comment (kevinrodrigues) [inline packages/components/pie-cookie-banner/test/component/pie-cookie-banner.spec.ts:79]: Same.

File: packages/components/pie-cookie-banner/test/component/pie-cookie-banner.spec.ts [hunk 6]
```diff
@@ -17,4 +41,112 @@ test.describe('PieCookieBanner - Component tests', () => {
...
+
+        await mount(PieCookieBanner, {
+            props: {} as CookieBannerProps,
+
+            on: {
+                [ON_COOKIE_BANNER_MANAGE_PREFS]: (event: Event) => events.push(event),
+            },
+        });
+
+        const cookieBanner = await page.locator(componentSelector);
```
Comment (kevinrodrigues) [inline packages/components/pie-cookie-banner/test/component/pie-cookie-banner.spec.ts:101]: Same and similar for the others too.

**Pull request merged**